src/matrix/room/Room.js

```diff
@@ -173,6 +174,7 @@ export class Room extends BaseRoom {
         if (Array.isArray(roomResponse.timeline?.events)) {
             removedPendingEvents = await this._sendQueue.removeRemoteEchos(roomResponse.timeline.events, txn, log);
         }
+        this._updatePowerLevels(roomResponse);
```

Comment:
This (e.g. updating in-memory data) should happen in afterSync, rather than writeSync, which is only about trying to persist the room response and returning data to incrementally update in-memory data with in afterSync. This is so that in-memory data structures are not altered until the storage transaction has successfully committed, so in case something goes wrong, we can safely retry without having data being out of sync between storage and in-memory. afterSync receives the return value from writeSync for this purpose. So here you'll actually want to return the new state events from the room response, and then call `_updatePowerLevels` with it in afterSync.

To find out about new state events, you'll need to look both at `roomResponse.state.events` (any state events since the last sync and the beginning of the returned timeline chunk in this sync response) and `roomResponse.timeline.events` (where state events are mixed with non-state events, and you can just filter events that have a state_key)